src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/SyntaxExtensions.cs

```diff
@@ -25,6 +25,31 @@ public static Location FindTypeExpressionOrNullLocation(this AttributeArgumentSy
             return walker.TypeExpressionLocation;
         }
 
+        public static AttributeData? FindAttributeData(this AttributeSyntax syntax, ISymbol targetSymbol)
+        {
+            AttributeTargetSpecifierSyntax attributeTarget = syntax.FirstAncestorOrSelf<AttributeListSyntax>().Target;
+            if (attributeTarget is not null)
+            {
+                switch (attributeTarget.Identifier.Kind())
+                {
+                    case SyntaxKind.ReturnKeyword:
+                        return ((IMethodSymbol)targetSymbol).GetReturnTypeAttributes().First(attributeSyntaxLocationMatches);
+                    case SyntaxKind.AssemblyKeyword:
+                        return targetSymbol.ContainingAssembly.GetAttributes().First(attributeSyntaxLocationMatches);
+                    case SyntaxKind.ModuleKeyword:
+                        return targetSymbol.ContainingModule.GetAttributes().First(attributeSyntaxLocationMatches);
+                    default:
+                        return null;
```

Comment:
I didn't look at the context of the callers or what's possible and what's not.

But there are other attribute targets to consider (only if they're relevant):

https://github.com/dotnet/roslyn/blob/1c480b07dbd55e2dce5580fc6a3e5b468d53c64d/src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_Locations.cs#L87-L97

Reply:
The rest of them aren't relevant enough to cover here. Returning `null` is fine. Technically, we don't need to handle `assembly` and `module` as well as this PR does, but I'd rather cleanly handle them than returning `null` if possible.

If your PR to add `IAttributeOperation` ever gets merged, we're going to switch to use that API from Roslyn and work with the `IOperation` tree instead of the syntax tree (and effectively obsolete all usages of this class).